lib/galaxy/tools/actions/__init__.py

```diff
@@ -236,7 +236,7 @@ def execute(self, tool, trans, incoming={}, return_job=False, set_output_hid=Tru
                 incoming[ "%s|__identifier__" % name ] = identifier
 
             for tag in data.tags:
-                if tag.user_tname == 'name':
+                if tag.user_tname == 'name' and tag.value not in [t.value for t in preserved_tags]:
```

Comment:
This only addresses regular datasets (LDA/HDA), right? I'm still seeing duplicated tags with collections, but I'm not exactly sure how/where collections and tags interact.